File: app/src/main/java/com/coinninja/coinkeeper/cn/wallet/service/CNServiceConnection.java
package com.coinninja.coinkeeper.cn.wallet.service;

import android.content.ComponentName;
import android.content.ServiceConnection;
import android.os.IBinder;

import com.coinninja.coinkeeper.cn.wallet.interfaces.CNWalletServicesInterface;
import com.coinninja.coinkeeper.util.Intents;
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil;

import javax.inject.Inject;

public class CNServiceConnection implements ServiceConnection {
    private CNWalletServicesInterface cnWalletServicesInterface;
    private LocalBroadCastUtil localBroadCastUtil;
    private boolean isBounded;

    @Inject
    public CNServiceConnection(LocalBroadCastUtil localBroadCastUtil) {
        this.localBroadCastUtil = localBroadCastUtil;
    }

    @Override
    public void onServiceConnected(ComponentName className, IBinder service) {
        CNWalletBinder binder = (CNWalletBinder) service;
        cnWalletServicesInterface = binder.getService();
        isBounded = true;
        localBroadCastUtil.sendBroadcast(Intents.ACTION_ON_SERVICE_CONNECTION_BOUNDED);
    }

    @Override
    public void onServiceDisconnected(ComponentName arg0) {
        isBounded = false;
    }

    public CNWalletServicesInterface getCNWalletServicesInterface() {
        return cnWalletServicesInterface;
    }

    public void setBounded(boolean bounded) {
        isBounded = bounded;
    }

    public boolean isBounded() {
        return isBounded;
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/cn/wallet/service/CNServiceConnection.kt
package com.coinninja.coinkeeper.cn.wallet.service

import android.content.ComponentName
import android.content.ServiceConnection
import android.os.IBinder
import app.dropbit.annotations.Mockable

import com.coinninja.coinkeeper.cn.wallet.interfaces.CNWalletServicesInterface
import com.coinninja.coinkeeper.util.DropbitIntents
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil

import javax.inject.Inject

@Mockable
class CNServiceConnection @Inject constructor(
        internal val localBroadCastUtil: LocalBroadCastUtil) : ServiceConnection {

    internal var cnWalletBinder: CNWalletBinder? = null

    val cnWalletServicesInterface: CNWalletServicesInterface? get() = cnWalletBinder?.service

    var isBounded: Boolean = false

    override fun onServiceConnected(className: ComponentName, service: IBinder) {
        cnWalletBinder = service as CNWalletBinder
        isBounded = true
        localBroadCastUtil.sendBroadcast(DropbitIntents.ACTION_ON_SERVICE_CONNECTION_BOUNDED)
    }

    override fun onServiceDisconnected(componentName: ComponentName) {
        isBounded = false
    }
}
